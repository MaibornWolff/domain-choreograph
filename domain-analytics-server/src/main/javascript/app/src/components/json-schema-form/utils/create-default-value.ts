import { JSONSchema4, JSONSchema4TypeName } from 'json-schema';

const defaultValues: Record<Exclude<JSONSchema4TypeName, 'object'>, any> = {
  string: undefined,
  number: undefined,
  integer: undefined,
  boolean: undefined,
  array: [],
  'null': null,
  any: ''
};

export function createDefaultValue(schema: JSONSchema4): any {
  const type = schema.type || 'any';
  if (type instanceof Array) {
    throw new Error('Array types are not supported');
  } else if (type === 'object') {
    return createDefaultValueForObject(schema);
  } else {
    return defaultValues[type];
  }
}

function createDefaultValueForObject(schema: JSONSchema4): Promise<object> {
  if (schema.type !== 'object') {
    throw new Error('Expect schema to be an object');
  }
  let defaultValue: any = {};
  if (schema.properties) {
    Object.entries(schema.properties).forEach(([key, property]) => {
      defaultValue[key] = createDefaultValue(property)
    });
  }
  return defaultValue;
}

