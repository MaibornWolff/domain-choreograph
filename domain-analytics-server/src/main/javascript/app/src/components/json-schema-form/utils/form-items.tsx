import { styled } from '~utils/styled';

export const FormItems = styled('div')`
    display: flex;
    flex-direction: column;

    & > * {
      margin-bottom: .5rem;
    }
`;
