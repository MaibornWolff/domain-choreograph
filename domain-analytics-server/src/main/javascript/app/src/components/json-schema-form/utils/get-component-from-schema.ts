import { FieldComponent } from 'clean-forms';
import { JSONSchema4, JSONSchema4TypeName } from 'json-schema';
import { DatePickerMsField } from '~components/json-schema-form/form-fields/date-picker-ms-field';
import { NumberTextField, TextField } from '~components/json-schema-form/form-fields/text-field';
import { JsonSchemaFormFieldProps } from '~components/json-schema-form/models/json-schema-form-field-props';

export interface JsonSchemaComponentMap {
  matcher?: { type: JSONSchema4TypeName, format?: string };
  field: FieldComponent<any, JsonSchemaFormFieldProps>;
}

const defaultMapping: JsonSchemaComponentMap[] = [
  { matcher: { type: 'integer', format: 'utc-millisec' }, field: DatePickerMsField },
  { matcher: { type: 'integer' }, field: NumberTextField },
  { matcher: { type: 'number' }, field: NumberTextField },
  { field: TextField }
];

export function getComponentFromSchema(
  schema: JSONSchema4,
  mapping: JsonSchemaComponentMap[] = defaultMapping
): FieldComponent<any, JsonSchemaFormFieldProps> {
  for (const map of mapping) {
    if (map.matcher == null) {
      return map.field;
    }
    const typeIsMatching = map.matcher.type === schema.type;
    const formatIsMatching = map.matcher.format == null || map.matcher.format === schema.format;
    if (typeIsMatching && formatIsMatching) {
      return map.field;
    }
  }
  throw new Error('Couldn\'t match the schema');
}
