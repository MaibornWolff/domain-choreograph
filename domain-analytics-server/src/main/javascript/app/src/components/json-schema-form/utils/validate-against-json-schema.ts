import { ValidationFunction } from 'clean-forms';
import { JSONSchema4 } from 'json-schema';
import { Validator } from 'jsonschema';


export function validateAgainstJsonSchema(schema: JSONSchema4): ValidationFunction {
  const validator = new Validator();
  return ({ value }) => {
    const result = validator.validate(value, schema as any);
    const errors = result.errors.map(error => {
      let message = error.message;
      let property = error.property
        .replace(/^instance\.?/, '');
      let schema = error.schema as JSONSchema4;
      if (error.name === 'required') {
        message = `is required`;
        schema = schema.properties![error.argument];
        property = joinProperties(property, error.argument);
      }

      const title = schema.title || '';
      message = `${title} ${message}`;

      return [ property, message ] as [string, string];
    });
    return errors;
  }
}

function joinProperties(path1: string, path2: string): string {
  if (path1 === '') {
    return path2;
  }
  if (path2 === '') {
    return path1;
  }
  return `${path1}.${path2}`;
}
