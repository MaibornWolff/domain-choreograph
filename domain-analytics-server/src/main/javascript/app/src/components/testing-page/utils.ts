export function escapeJavaClass(javaClass: string) {
  return javaClass.replace(/\./g, '-');
}

export function reverseJavaClassEscape(javaClass: string) {
  return javaClass.replace(/-/g, '.');
}
