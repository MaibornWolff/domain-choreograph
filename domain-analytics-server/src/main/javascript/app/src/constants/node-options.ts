export const NODE_OPTIONS = {
  minHeightInPx: 50,
  maxWidthInPx: 250,
  sideMarginInPx: 16,
};

export type NodeOptions = typeof NODE_OPTIONS;
