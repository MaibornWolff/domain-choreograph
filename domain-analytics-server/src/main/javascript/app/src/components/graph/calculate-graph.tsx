import { line } from 'd3-shape';
import * as dagre from 'dagre-layout';
import memoize from 'fast-memoize';
import { EdgeModel } from '~components/graph/edge/edge';
import { NodeModel } from '~components/graph/node/node';
import { NodeOptions } from '~constants/node-options';
import { ChoreoDependency, ChoreoGraph } from '~types/choreo-graph';
import { selectPreviewTextFromNode } from '~utils/node.utils';

export interface VisualGraph {
  nodes: NodeModel[];
  edges: EdgeModel[];
  scale: number;
  height: number | undefined;
  width: number | undefined;
}

export interface CalculateGraphProps {
  graph: ChoreoGraph;
  scopeKey: keyof ChoreoGraph['scopes'];
  nodeOptions: NodeOptions;
  targetWidth?: number;
  targetHeight?: number;
}

export const calculateGraph = memoize(_calculateGraph);

function _calculateGraph({ graph, scopeKey, nodeOptions, targetWidth, targetHeight }: CalculateGraphProps): VisualGraph {
  const scope = graph.scopes[scopeKey];

  const isDependencyRelevant = (dependency: ChoreoDependency) => scope.nodes.includes(dependency.target);
  const edgeData = graph.dependencies.filter(isDependencyRelevant);
  const nodeIds = edgeData
    .map(e => [e.src, e.target])
    .reduce((result, ids) => [...result, ...ids], []);
  const nodeData = nodeIds.map(id => graph.nodes[id]);

  const g = new dagre.graphlib.Graph();

  g.setGraph({});
  g.setDefaultEdgeLabel(() => ({}));

  const isOutsideScope = (nodeId: string) => !scope.nodes.includes(nodeId);
  nodeData.forEach(n => {
      const previewText = selectPreviewTextFromNode(n);
      g.setNode(n.id, {
        ...n,
        outsideScope: isOutsideScope(n.id),
        executionContext: graph.executionContexts[n.id],
        width: calculateNodeTextWidthInPx(n.name, previewText, nodeOptions),
        height: calculateNodeTextHeightInPx(previewText, nodeOptions),
      });
    }
  );
  edgeData.map(e =>
    g.setEdge(e.src, e.target, {
      id: createEdgeId(e),
      outsideScope: isOutsideScope(e.src) || isOutsideScope(e.target)
    })
  );

  dagre.layout(g);

  interface Point {
    x: number;
    y: number;
  }

  let innerHeight = g.graph().height || 0;
  let innerWidth = g.graph().width || 0;
  innerHeight = isFinite(innerHeight) ? innerHeight : 0;
  innerWidth = isFinite(innerWidth) ? innerWidth : 0;
  const scale = calculateScale(
    { width: targetWidth, height: targetHeight },
    { width: innerWidth, height: innerHeight }
  );

  const createEdgeLine = line<Point>()
    .x(p => p.x * scale)
    .y(p => p.y * scale)
  ;
  const nodes: NodeModel[] = g.nodes()
    .map(id => g.node(id));
  const edges: EdgeModel[] = g.edges()
    .map(id => g.edge(id))
    .map(edge => ({
      ...edge,
      line: createEdgeLine(edge.points)
    }));

  return {
    nodes,
    edges,
    height: innerHeight * scale,
    width: innerWidth * scale * 1.1,
    scale,
  };
}

interface WidthHeight {
  width?: number;
  height?: number;
}

export function createEdgeId({src, target}: ChoreoDependency): string {
  return `${src}->${target}`;
}

function calculateScale(target: WidthHeight, origin: WidthHeight): number {
  let result = 1;
  if (target.width && origin.width) {
    result = Math.min(target.width / origin.width, result);
  }
  if (target.height && origin.height) {
    result = Math.min(target.height / origin.height, result);
  }
  return result;
}

function calculateNodeTextWidthInPx(name: string, value: string | undefined, nodeOptions: NodeOptions): number {
  const { width: divWidth } = calculateClientRect(name + '<br>' + value);
  const { width: preWidth } = calculateClientRect(name + '<br>' + value, { elementType: 'pre' });
  return Math.min(Math.max(divWidth, preWidth), nodeOptions.maxWidthInPx) + nodeOptions.sideMarginInPx * 2;
}

function calculateNodeTextHeightInPx(value: string | undefined, nodeOptions: NodeOptions): number {
  if (value == null) {
    return nodeOptions.minHeightInPx;
  }

  const options: CalculateClientRectOptions = {
    fontSizeInRem: .8,
    elementType: 'pre',
    maxWidth: nodeOptions.maxWidthInPx,
  };
  const { height } = calculateClientRect(value, options);
  return nodeOptions.minHeightInPx + height;
}

interface CalculateClientRectOptions {
  fontSizeInRem?: number;
  elementType?: string;
  maxWidth?: number;
}

function calculateClientRect(text: string, {
  fontSizeInRem = 1,
  elementType = 'div',
  maxWidth,
}: CalculateClientRectOptions = {}): ClientRect {
  const element = document.createElement(elementType);
  element.style.display = 'inline-block';
  element.style.fontSize = `${fontSizeInRem}rem`;
  element.style.whiteSpace = 'pre-wrap';
  element.style.padding = '0 8px';
  element.innerHTML = text;
  if (maxWidth) {
    element.style.maxWidth = `${maxWidth}px`;
  }
  document.body.appendChild(element);
  const rect = element.getBoundingClientRect();
  element.remove();
  return rect;
}
