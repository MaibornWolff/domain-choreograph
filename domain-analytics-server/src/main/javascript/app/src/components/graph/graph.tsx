import * as React from 'react';
import { CSSProperties } from 'react';
import { calculateGraph } from '~components/graph/calculate-graph';
import { Edge } from '~components/graph/edge/edge';
import { getSelectedOptionsOrUndefined, SelectedOptions } from '~components/graph/models/selected-options';
import { Node } from '~components/graph/node/node';
import { NodeOptions } from '~constants/node-options';
import { ChoreoGraph } from '~types/choreo-graph';
import { styled } from '~utils/styled';

export const EDGE_ARROW_WIDTH = 7;
export const EDGE_ARROW_HEIGHT = 8;

const NodeWrapper = styled('div')`
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
`;

const EdgeWrapper = NodeWrapper.withComponent('svg');

export interface GraphProps {
  className?: string;
  scope: string;
  graph: ChoreoGraph | null;
  nodeOptions: NodeOptions;
  width?: number;
  height?: number;
  toggleNode?: (nodeId: string) => void;
  selectedNodeIds?: Record<string, boolean | SelectedOptions>;
  selectedEdgeIds?: Record<string, boolean | SelectedOptions>;
}

export interface GraphState {
}

export class Graph extends React.PureComponent<GraphProps, GraphState> {
  public render() {
    const props = this.props;
    const {
      graph: choreoGraph,
      scope: scopeKey,
      nodeOptions,
      toggleNode,
      selectedNodeIds = {},
      selectedEdgeIds = {}
    } = props;
    if (!choreoGraph) {
      return <p>No Data</p>;
    }

    const { width, height, nodes, edges, scale } = calculateGraph({
      graph: choreoGraph,
      nodeOptions,
      scopeKey,
      targetHeight: props.height,
      targetWidth: props.width,
    });

    const widthHeight: CSSProperties = { height, width };

    const style: CSSProperties = {
      ...widthHeight,
      position: 'relative',
      margin: 'auto'
    };

    return (
      <div style={style} className={this.props.className || ''}>
        <NodeWrapper>
          {nodes.map((node, i) => <Node
            toggleNode={toggleNode}
            isSelected={!!selectedNodeIds[node.id]}
            selectedOptions={getSelectedOptionsOrUndefined(selectedNodeIds[node.id])}
            key={i}
            node={node}
            scale={scale}
            options={nodeOptions}/>
          )}
        </NodeWrapper>
        <EdgeWrapper
          style={widthHeight}
        >
          {edges.map((edge, i) => <Edge
            key={i}
            edge={edge}
            selectedOptions={getSelectedOptionsOrUndefined(selectedEdgeIds[edge.id])}
            isSelected={!!selectedEdgeIds[edge.id]}
          />)}
        </EdgeWrapper>
      </div>
    );
  }
}
