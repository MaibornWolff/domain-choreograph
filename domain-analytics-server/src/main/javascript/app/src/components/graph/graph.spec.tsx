import * as React from 'react';
import { NODE_OPTIONS } from '~constants/node-options';
import { graphMock } from '~types/cheoreo-graph.mock';
import { ChoreoGraph } from '~types/choreo-graph';
import { renderWithProviders } from '~utils/test.utils';
import { Graph } from './graph';

describe('Graph', () => {
  it('should render without data', () => {
    expect(renderWithProviders(<Graph graph={null} scope={'root'} nodeOptions={NODE_OPTIONS}/>)).toMatchSnapshot();
  });

  it('should render with data', () => {
    expect(renderWithProviders(<Graph graph={graphMock} scope={'root'} nodeOptions={NODE_OPTIONS}/>)).toMatchSnapshot();
  });

  it('should render with just one node', () => {
    const graph: ChoreoGraph = {
      nodes: {
        A: {
          id: 'A',
          name: 'A',
          javaClass: 'A',
          scope: 'root',
          value: {}
        }
      },
      dependencies: [],
      scopes: {
        root: {
          id: 'root',
          executionContext: 'Application',
          nodes: [
            'A',
            'F'
          ]
        },
      },
      executionContexts: {
        Application: {
          id: 'Application',
          scopes: [
            'root'
          ]
        },
      }
    };

    expect(renderWithProviders(<Graph graph={graph} scope={'root'} nodeOptions={NODE_OPTIONS}/>)).toMatchSnapshot();
  });

  it('should select edges', () => {
    expect(renderWithProviders(<Graph
      graph={graphMock}
      scope={'root'}
      nodeOptions={NODE_OPTIONS}
      selectedEdgeIds={{ '0->1': true }}
    />)).toMatchSnapshot();
  });

  it('should select edges with color', () => {
    expect(renderWithProviders(<Graph
      graph={graphMock}
      scope={'root'}
      nodeOptions={NODE_OPTIONS}
      selectedEdgeIds={{ '0->1': { color: 'red' } }}
    />)).toMatchSnapshot();
  });

  it('should select nodes', () => {
    expect(renderWithProviders(<Graph
      graph={graphMock}
      scope={'root'}
      nodeOptions={NODE_OPTIONS}
      selectedNodeIds={{ 0: true }}
    />)).toMatchSnapshot();
  });

  it('should select nodes with color', () => {
    expect(renderWithProviders(<Graph
      graph={graphMock}
      scope={'root'}
      nodeOptions={NODE_OPTIONS}
      selectedNodeIds={{ 0: { color: 'red' } }}
    />)).toMatchSnapshot();
  });
});
