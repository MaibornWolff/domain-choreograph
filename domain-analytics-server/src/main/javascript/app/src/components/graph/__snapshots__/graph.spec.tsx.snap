// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Graph should render with data 1`] = `
.c5 {
  fill: none;
  stroke: #999;
  stroke-width: 2px;
  -webkit-transition: all 0.3s cubic-bezier(.25,.8,.25,1);
  transition: all 0.3s cubic-bezier(.25,.8,.25,1);
}

.c6 {
  fill: #999;
}

.c1 {
  position: absolute;
  background: #FFFFFF;
  border: solid 2px rgba(0,0,0,0.3);
  box-shadow: 0 0 5px 1px rgba(0,0,0,0.15);
  border-radius: 5px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 0 16px;
  width: 32px;
  height: 50px;
  font-size: calc(1rem * 1);
  box-sizing: border-box;
  left: 100px;
  top: 0px;
  cursor: pointer;
  -webkit-transition: all 0.3s cubic-bezier(.25,.8,.25,1);
  transition: all 0.3s cubic-bezier(.25,.8,.25,1);
  z-index: 1;
}

.c3 {
  position: absolute;
  background: #FFFFFF;
  border: solid 2px rgba(0,0,0,0.3);
  box-shadow: 0 0 5px 1px rgba(0,0,0,0.15);
  border-radius: 5px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 0 16px;
  width: 32px;
  height: 50px;
  font-size: calc(1rem * 1);
  box-sizing: border-box;
  left: 100px;
  top: 100px;
  cursor: pointer;
  -webkit-transition: all 0.3s cubic-bezier(.25,.8,.25,1);
  transition: all 0.3s cubic-bezier(.25,.8,.25,1);
  z-index: 1;
}

.c2 {
  overflow: hidden;
  text-overflow: ellipsis;
  max-width: 100%;
}

.c0 {
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
}

.c4 {
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
}

<div
  class=""
  style="height:150px;width:255.20000000000002px;position:relative;margin:auto"
>
  <div
    class="c0"
  >
    <div
      class="c1"
      scale="1"
    >
      <div
        class="c2"
      >
        node-0
      </div>
    </div>
    <div
      class="c3"
      scale="1"
    >
      <div
        class="c2"
      >
        node-1
      </div>
    </div>
  </div>
  <svg
    class="c4"
    style="height:150px;width:255.20000000000002px"
  >
    <g>
      <path
        class="c5"
        d="M116,50L116,75L116,100"
      />
      <polygon
        class="c6"
        height="8"
        points="116,100 112.5,92 119.5,92"
        transform="rotate(0 116 100)"
        width="7"
        x="116"
        y="100"
      />
    </g>
  </svg>
</div>
`;

exports[`Graph should render with just one node 1`] = `
.c0 {
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
}

.c1 {
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
}

<div
  class=""
  style="height:0;width:0;position:relative;margin:auto"
>
  <div
    class="c0"
  />
  <svg
    class="c1"
    style="height:0;width:0"
  />
</div>
`;

exports[`Graph should render without data 1`] = `
<p>
  No Data
</p>
`;

exports[`Graph should select edges 1`] = `
.c5 {
  fill: none;
  stroke: #657fcc;
  stroke-width: 2px;
  -webkit-transition: all 0.3s cubic-bezier(.25,.8,.25,1);
  transition: all 0.3s cubic-bezier(.25,.8,.25,1);
}

.c6 {
  fill: #657fcc;
}

.c1 {
  position: absolute;
  background: #FFFFFF;
  border: solid 2px rgba(0,0,0,0.3);
  box-shadow: 0 0 5px 1px rgba(0,0,0,0.15);
  border-radius: 5px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 0 16px;
  width: 32px;
  height: 50px;
  font-size: calc(1rem * 1);
  box-sizing: border-box;
  left: 100px;
  top: 0px;
  cursor: pointer;
  -webkit-transition: all 0.3s cubic-bezier(.25,.8,.25,1);
  transition: all 0.3s cubic-bezier(.25,.8,.25,1);
  z-index: 1;
}

.c3 {
  position: absolute;
  background: #FFFFFF;
  border: solid 2px rgba(0,0,0,0.3);
  box-shadow: 0 0 5px 1px rgba(0,0,0,0.15);
  border-radius: 5px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 0 16px;
  width: 32px;
  height: 50px;
  font-size: calc(1rem * 1);
  box-sizing: border-box;
  left: 100px;
  top: 100px;
  cursor: pointer;
  -webkit-transition: all 0.3s cubic-bezier(.25,.8,.25,1);
  transition: all 0.3s cubic-bezier(.25,.8,.25,1);
  z-index: 1;
}

.c2 {
  overflow: hidden;
  text-overflow: ellipsis;
  max-width: 100%;
}

.c0 {
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
}

.c4 {
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
}

<div
  class=""
  style="height:150px;width:255.20000000000002px;position:relative;margin:auto"
>
  <div
    class="c0"
  >
    <div
      class="c1"
      scale="1"
    >
      <div
        class="c2"
      >
        node-0
      </div>
    </div>
    <div
      class="c3"
      scale="1"
    >
      <div
        class="c2"
      >
        node-1
      </div>
    </div>
  </div>
  <svg
    class="c4"
    style="height:150px;width:255.20000000000002px"
  >
    <g>
      <path
        class="c5"
        d="M116,50L116,75L116,100"
      />
      <polygon
        class="c6"
        height="8"
        points="116,100 112.5,92 119.5,92"
        transform="rotate(0 116 100)"
        width="7"
        x="116"
        y="100"
      />
    </g>
  </svg>
</div>
`;

exports[`Graph should select edges with color 1`] = `
.c5 {
  fill: none;
  stroke: red;
  stroke-width: 2px;
  -webkit-transition: all 0.3s cubic-bezier(.25,.8,.25,1);
  transition: all 0.3s cubic-bezier(.25,.8,.25,1);
}

.c6 {
  fill: red;
}

.c1 {
  position: absolute;
  background: #FFFFFF;
  border: solid 2px rgba(0,0,0,0.3);
  box-shadow: 0 0 5px 1px rgba(0,0,0,0.15);
  border-radius: 5px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 0 16px;
  width: 32px;
  height: 50px;
  font-size: calc(1rem * 1);
  box-sizing: border-box;
  left: 100px;
  top: 0px;
  cursor: pointer;
  -webkit-transition: all 0.3s cubic-bezier(.25,.8,.25,1);
  transition: all 0.3s cubic-bezier(.25,.8,.25,1);
  z-index: 1;
}

.c3 {
  position: absolute;
  background: #FFFFFF;
  border: solid 2px rgba(0,0,0,0.3);
  box-shadow: 0 0 5px 1px rgba(0,0,0,0.15);
  border-radius: 5px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 0 16px;
  width: 32px;
  height: 50px;
  font-size: calc(1rem * 1);
  box-sizing: border-box;
  left: 100px;
  top: 100px;
  cursor: pointer;
  -webkit-transition: all 0.3s cubic-bezier(.25,.8,.25,1);
  transition: all 0.3s cubic-bezier(.25,.8,.25,1);
  z-index: 1;
}

.c2 {
  overflow: hidden;
  text-overflow: ellipsis;
  max-width: 100%;
}

.c0 {
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
}

.c4 {
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
}

<div
  class=""
  style="height:150px;width:255.20000000000002px;position:relative;margin:auto"
>
  <div
    class="c0"
  >
    <div
      class="c1"
      scale="1"
    >
      <div
        class="c2"
      >
        node-0
      </div>
    </div>
    <div
      class="c3"
      scale="1"
    >
      <div
        class="c2"
      >
        node-1
      </div>
    </div>
  </div>
  <svg
    class="c4"
    style="height:150px;width:255.20000000000002px"
  >
    <g>
      <path
        class="c5"
        d="M116,50L116,75L116,100"
      />
      <polygon
        class="c6"
        height="8"
        points="116,100 112.5,92 119.5,92"
        transform="rotate(0 116 100)"
        width="7"
        x="116"
        y="100"
      />
    </g>
  </svg>
</div>
`;

exports[`Graph should select nodes 1`] = `
.c5 {
  fill: none;
  stroke: #999;
  stroke-width: 2px;
  -webkit-transition: all 0.3s cubic-bezier(.25,.8,.25,1);
  transition: all 0.3s cubic-bezier(.25,.8,.25,1);
}

.c6 {
  fill: #999;
}

.c3 {
  position: absolute;
  background: #FFFFFF;
  border: solid 2px rgba(0,0,0,0.3);
  box-shadow: 0 0 5px 1px rgba(0,0,0,0.15);
  border-radius: 5px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 0 16px;
  width: 32px;
  height: 50px;
  font-size: calc(1rem * 1);
  box-sizing: border-box;
  left: 100px;
  top: 100px;
  cursor: pointer;
  -webkit-transition: all 0.3s cubic-bezier(.25,.8,.25,1);
  transition: all 0.3s cubic-bezier(.25,.8,.25,1);
  z-index: 1;
}

.c1 {
  position: absolute;
  background: #FFFFFF;
  border: solid 2px #3956ac;
  box-shadow: 0 0 5px 1px rgba(57,86,172,0.5);
  border-radius: 5px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 0 16px;
  width: 32px;
  height: 50px;
  font-size: calc(1rem * 1);
  box-sizing: border-box;
  left: 100px;
  top: 0px;
  cursor: pointer;
  -webkit-transition: all 0.3s cubic-bezier(.25,.8,.25,1);
  transition: all 0.3s cubic-bezier(.25,.8,.25,1);
  z-index: 1;
}

.c2 {
  overflow: hidden;
  text-overflow: ellipsis;
  max-width: 100%;
}

.c0 {
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
}

.c4 {
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
}

<div
  class=""
  style="height:150px;width:255.20000000000002px;position:relative;margin:auto"
>
  <div
    class="c0"
  >
    <div
      class="c1"
      scale="1"
    >
      <div
        class="c2"
      >
        node-0
      </div>
    </div>
    <div
      class="c3"
      scale="1"
    >
      <div
        class="c2"
      >
        node-1
      </div>
    </div>
  </div>
  <svg
    class="c4"
    style="height:150px;width:255.20000000000002px"
  >
    <g>
      <path
        class="c5"
        d="M116,50L116,75L116,100"
      />
      <polygon
        class="c6"
        height="8"
        points="116,100 112.5,92 119.5,92"
        transform="rotate(0 116 100)"
        width="7"
        x="116"
        y="100"
      />
    </g>
  </svg>
</div>
`;

exports[`Graph should select nodes with color 1`] = `
.c5 {
  fill: none;
  stroke: #999;
  stroke-width: 2px;
  -webkit-transition: all 0.3s cubic-bezier(.25,.8,.25,1);
  transition: all 0.3s cubic-bezier(.25,.8,.25,1);
}

.c6 {
  fill: #999;
}

.c3 {
  position: absolute;
  background: #FFFFFF;
  border: solid 2px rgba(0,0,0,0.3);
  box-shadow: 0 0 5px 1px rgba(0,0,0,0.15);
  border-radius: 5px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 0 16px;
  width: 32px;
  height: 50px;
  font-size: calc(1rem * 1);
  box-sizing: border-box;
  left: 100px;
  top: 100px;
  cursor: pointer;
  -webkit-transition: all 0.3s cubic-bezier(.25,.8,.25,1);
  transition: all 0.3s cubic-bezier(.25,.8,.25,1);
  z-index: 1;
}

.c1 {
  position: absolute;
  background: #FFFFFF;
  border: solid 2px red;
  box-shadow: 0 0 5px 1px rgba(255,0,0,0.5);
  border-radius: 5px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 0 16px;
  width: 32px;
  height: 50px;
  font-size: calc(1rem * 1);
  box-sizing: border-box;
  left: 100px;
  top: 0px;
  cursor: pointer;
  -webkit-transition: all 0.3s cubic-bezier(.25,.8,.25,1);
  transition: all 0.3s cubic-bezier(.25,.8,.25,1);
  z-index: 1;
}

.c2 {
  overflow: hidden;
  text-overflow: ellipsis;
  max-width: 100%;
}

.c0 {
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
}

.c4 {
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
}

<div
  class=""
  style="height:150px;width:255.20000000000002px;position:relative;margin:auto"
>
  <div
    class="c0"
  >
    <div
      class="c1"
      scale="1"
    >
      <div
        class="c2"
      >
        node-0
      </div>
    </div>
    <div
      class="c3"
      scale="1"
    >
      <div
        class="c2"
      >
        node-1
      </div>
    </div>
  </div>
  <svg
    class="c4"
    style="height:150px;width:255.20000000000002px"
  >
    <g>
      <path
        class="c5"
        d="M116,50L116,75L116,100"
      />
      <polygon
        class="c6"
        height="8"
        points="116,100 112.5,92 119.5,92"
        transform="rotate(0 116 100)"
        width="7"
        x="116"
        y="100"
      />
    </g>
  </svg>
</div>
`;
